### PR TITLE
[zify] add support for Nat.div, Nat.mod, Nat.pow

### LINKED ISSUE #11447: zify should support Nat.div and Nat.modulo
#### Description of the problem

With `PreOmega.Z.to_euclidean_division_equations`, tactics like `lia` can handle `/` and `mod` in `Z` and `N`.  However, `zify` handles neither `/` nor `mod` on `nat`.  This should be corrected.

cc @fajb 

#### Coq Version

8.11

### PR DESCRIPTION
**Pull request by mrhaandi**
The required declarations `BinOp Nat.modulo`, `BinOp Nat.div`, `BinOp Nat.pow` are added to `ZifyInst`.
Also `Saturate Z.div`, `Saturate Z.modulo` are added.

I did not follow the suggestion https://github.com/coq/coq/issues/11447#issuecomment-745144254 of using ad-hoc lemmas from `Zdiv.v` (because of circular dependencies).
Instead, I added missing `inj_*` lemmas to `Nnat.v`, which were sufficient.

Additionally, this PR cleans up `Zdiv.v` a little (no more `PeanoNat`) and renders `div_Zdiv` and `mod_Zmod` as derived from `inj_*`.
Actually, I would prefer to remove `div_Zdiv` and `mod_Zmod` completely.

Here is a demo snippet which uses the added functionality:
```
Require Import ZArith Lia.

Ltac Zify.zify_post_hook ::= Z.to_euclidean_division_equations.
 
Goal forall (n:nat), (n * 3) mod 3 = 0.
Proof. lia. Qed.
 
Goal forall (n:nat), (n * 3) / 3 = n.
Proof. lia. Qed.
 
(* not sure how to fix *)
Goal forall (m n:nat), m > 0 -> (n * m) mod m = 0.
Proof. Fail nia. Abort.
 
Goal forall (m n:nat), m > 0 -> (n * m) / m = n.
Proof. nia. Qed.
```
Does anyone know how to fix the failing third Goal?

**Kind:** feature.

Closes #11447

Requires https://github.com/coq/coq/pull/14045, https://github.com/coq/coq/pull/14086

Associted PRs:
- [x] math-classes (merged): https://github.com/coq-community/math-classes/pull/98
- [x] flocq: (merged) https://github.com/mrhaandi/flocq/tree/N-Z-div-mod
@silene diff: [flocq.diff.txt](https://github.com/coq/coq/files/6521863/flocq.diff.txt)

- [x] Added / updated test-suite

- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).
- [x] Overlay pull requests (if this breaks 3rd party developments in CI, see
https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md for details)

### CODE REVIEW

**Comment by JasonGross** [comment]
> Does anyone know how to fix the failing third Goal?

`nia` is not complete, and this goal already fails in `Z`.  In general I've found that I have to `cut` with an equation about the quotient in order to handle goals involving the remainder:
```coq
Require Import Lia ZArith.
Ltac Zify.zify_post_hook ::= Z.to_euclidean_division_equations.
Local Open Scope Z_scope.
Goal forall (m n:Z), 0 <= n -> 0 < m -> (n * m) mod m = 0.
Proof.
  intros.
  Fail nia.
  cut ((n * m) / m = n); nia.
Qed.
```

**Comment by mrhaandi** [comment]
CI shows the following regression.
Previously the following works, whereas now it breaks:
```
Goal forall n, n mod 2 < 2 -> n mod 2 = 0 \/ n mod 2 = 1.
Proof. lia. Qed.
```
It can be again made working adding `Ltac Zify.zify_post_hook ::= Z.to_euclidean_division_equations.`.
Then, the assumption `n mod 2 < 2` is not even needed. However, the regression breaks existing code that does not (possibly on purpose) add the hook.
`mod` is translated to `Z` and then `lia` gets stuck without the hook. Suggestions?

**Comment by JasonGross** [comment]
> Suggestions?

I think this is an incompleteness in the part of `zify` written in OCaml.  In particular, the following instances should be enough, but they are not:

```coq
Require Import Lia ZArith Zify ZifyInst ZifyClasses.
Lemma Z_div_pos_full a b : (0 <= a -> 0 <= b -> 0 <= a / b)%Z.
Proof. destruct b; intros; now (rewrite Zdiv_0_r + apply Z.div_pos). Qed.
Instance SatDiv : Saturate Z.div :=
  {|
  PArg1 := fun x => 0 <= x;
  PArg2 := fun y => 0 <= y;
  PRes  := fun r => 0 <= r;
  SatOk := Z_div_pos_full
  |}%Z.
Add Zify Saturate SatDiv.
Lemma Z_mod_nonneg_full a b : (0 <= b -> 0 <= a mod b)%Z.
Proof. destruct b; intros; now (rewrite Zmod_0_r + apply Z.mod_pos_bound). Qed.
Instance SatMod : Saturate Z.modulo :=
  {|
  PArg1 := fun x => True;
  PArg2 := fun y => 0 <= y;
  PRes  := fun r => 0 <= r;
  SatOk := fun a b _ => @Z_mod_nonneg_full a b
  |}%Z.
Add Zify Saturate SatMod.
Axiom admit : forall {T}, T.
Instance Op_mod : BinOp Nat.modulo
  := {| TBOp := Z.modulo ; TBOpInj := admit |}.
Add Zify BinOp Op_mod.
Instance Op_div : BinOp Nat.div :=
  {| TBOp := Z.div ; TBOpInj := admit |}.
Add Zify BinOp Op_div.
Goal forall n, n mod 2 < 2 -> n mod 2 = 0 \/ n mod 2 = 1.
Proof. assert (0 <= 2) by lia. pose proof I. zify. lia. Qed.
```

There are two issues, both of which I think need to be fixed in the OCaml code (cc @fajb):
1. The `Saturate` instances do not handle the case where one of the arguments can be unconstrained (requiring me to `pose proof I`)
2. The saturation procedure does not handle the case where the input constraints are true because the inputs are constants.

**Comment by JasonGross** [comment]
> CI shows the following regression.
> Previously the following works, whereas now it breaks:

@mrhaandi Please add this test-case to the test-suite; I think ideally anytime a CI failure reveals something not caught by the test-suite, we should update the test-suite to cover that case, if it's not too hard.

**Comment by fajb** [comment]
@mrhaandi @JasonGross Thanks for the PR and the investigation. 
The `Saturate` thingy   has indeed limitations: it performs a syntactic matching in the context.
I'll look into it - it looks feasible to "invent" positivity proofs for positive constants.
 In the meantime, this is unfortunately blocking this PR...
A quick workaround would be to place the Nat.div and Nat.modulo in a separate file.
Not quite satisfactory but at least this would not break existing code.

**Comment by mrhaandi** [comment]
I looked at all the CI problems.
All (except for iris) are exactly due to the regression out earlier.
However, they are easy to fix locally by simply `generalize (x mod y)` the particular expressions that would be decomposed by `lia`.
Of course, fixing the actual regression would be better.
> A quick workaround would be to place the Nat.div and Nat.modulo in a separate file.

This is also a quick (but less preferable) option.

iris breaks for a different reason and is easy to fix by better code (I'll create a PR).

**Comment by RalfJung** [comment]
> Edit: @RalfJung Hmmm, I have hard on gitlab trying to fork iris_examples.

Yeah, see [our docs for that](https://gitlab.mpi-sws.org/iris/iris/-/blob/master/CONTRIBUTING.md#how-to-submit-a-merge-request).

But I also just submitted your change to the repo, so no need to go through that process. :)

**Comment by JasonGross** [comment]
I think it would be better to find a way to maintain backwards compatibility than to make `lia` weaker.  My preference would indeed be to stick the `Add Zify BinOp` declarations in a separate file (maybe also put the ones about `Saturate` that I suggested there), and then we can have a future PR that merges them back into the main file, which is to be merged only after the saturation code is improved.

**Comment by mrhaandi** [comment]
> I think it would be better to find a way to maintain backwards compatibility than to make `lia` weaker. My preference would indeed be to stick the `Add Zify BinOp` declarations in a separate file (maybe also put the ones about `Saturate` that I suggested there), and then we can have a future PR that merges them back into the main file, which is to be merged only after the saturation code is improved.

I now added suggested `Saturate` instances and test cases to `test-suite` containing regressions which do currently fail.
Hopefully those can be resolved by @fajb's efforts. If it turns out to be too difficult, then I would refactor the PR into a module e.g. `ZifyNatExtra`.

**Comment by fajb** [comment]
@mrhaandi PR #14045 may solve the regression. 
I made a quick test on top of your PR. I am hopeful that the regressions will be due to `zify/lia` being stronger.

**Comment by mrhaandi** [comment]
> @mrhaandi PR #14045 may solve the regression.
> I made a quick test on top of your PR. I am hopeful that the regressions will be due to `zify/lia` being stronger.

I noticed that you removed `Saturate Z.mul`. Are the newly added `Saturate Z.div`, `Saturate Z.modulo` still ok?

**Comment by fajb** [comment]
@mrhaandi `Saturate Z.div`, `Saturate Z.modulo` should be ok.

**Comment by mrhaandi** [comment]
> @mrhaandi PR #14045 may solve the regression.
> I made a quick test on top of your PR. I am hopeful that the regressions will be due to `zify/lia` being stronger.

Yes, it seems that any issues in ci due to the previous regression are gone in the PR combination.
There are several new errors, all of which because of `zify/lia` being stronger.

**Comment by mrhaandi** [comment]
@fajb I conservatively updated `Int63.v` which previously broke upon rebase of https://github.com/coq/coq/pull/14045 upon this.
Now the combination of both PRs builds and passes regressions tests.

File: theories/micromega/ZifyInst.v
Comment (JasonGross) [inline theories/micromega/ZifyInst.v:534]: You probably don't need `%Z` here

File: theories/micromega/ZifyInst.v
Comment (JasonGross) [inline theories/micromega/ZifyInst.v:543]: You probably don't need `%Z` here

File: theories/ZArith/Zdiv.v
Comment (JasonGross) [inline theories/ZArith/Zdiv.v:732]: Perhaps this should be an actual deprecation like
https://github.com/coq/coq/blob/afc828b3e207dd39c59d1501d570a88b2012fd2c/theories/Init/Numeral.v#L15

File: theories/ZArith/Zdiv.v
Comment (JasonGross) [inline theories/ZArith/Zdiv.v:737]: Perhaps this should be an actual deprecation like
https://github.com/coq/coq/blob/afc828b3e207dd39c59d1501d570a88b2012fd2c/theories/Init/Numeral.v#L15

File: theories/micromega/ZifyInst.v
Comment (mrhaandi) [reply theories/micromega/ZifyInst.v]: fixed

File: theories/micromega/ZifyInst.v
Comment (mrhaandi) [reply theories/micromega/ZifyInst.v]: fixed

File: theories/ZArith/Zdiv.v
Comment (mrhaandi) [reply theories/ZArith/Zdiv.v]: fixed. However, if (in an other lemma) I `apply` the deprecated lemma, I get no deprecation warning.
The [documentation](https://coq.inria.fr/distrib/current/refman/using/libraries/writing.html?highlight=deprecated#coq:attr.deprecated) says 
> This attribute is supported by the following commands: Ltac, Tactic Notation, Notation, Infix.

So I assumed it would not work for lemmas (and therefore I initially wrote a comment).

File: theories/ZArith/Zdiv.v
Comment (mrhaandi) [reply theories/ZArith/Zdiv.v]: fixed; also see above.

File: theories/ZArith/Zdiv.v
Comment (mrhaandi) [reply theories/ZArith/Zdiv.v]: related: https://github.com/coq/coq/issues/12266

File: theories/NArith/Nnat.v
Comment (silene) [inline theories/NArith/Nnat.v:231]: Presumably, this hypothesis will become useless once #14086 is merged, right? If so, I suggest we wait for it.

Comment (mrhaandi) [reply theories/NArith/Nnat.v:231]: Yes.

**Comment by mrhaandi** [comment]
@fajb after https://github.com/coq/coq/pull/14045, what is the status of options such as
```
Ltac Zify.zify_post_hook ::= Z.to_euclidean_division_equations. needs to be replaced by
Ltac zify_convert_to_euclidean_division_equations_flag ::= constr:(true).
```

**Comment by fajb** [comment]
If you wish to totally eliminate divisions, you need to set one or the other.
The `Saturate` rules are applied after the potential elimination of division.

File: theories/ZArith/Zdiv.v
Comment (Blaisorblade) [reply theories/ZArith/Zdiv.v]: You can still deprecate lemmas today (https://coq.discourse.group/t/deprecated/814/2?u=blaisorblade) by wrapping them in notations. But you'd need to pick a name for the wrapped lemma.

**Comment by JasonGross** [comment]
@coqbot CI minimize

**Comment by mrhaandi** [comment]
@fajb As https://github.com/coq/coq/pull/14086 is now merged, `zify` can be simplified.
For example, consider `Goal forall (n:Z), (n mod n = 0)%Z.`. `zify` creates the assumption `n <> 0%Z -> n = (n * q + r)%Z`, where `n <> 0` is because of the old `mod` definition. Now, one can use the new `Z_div_mod_eq_full` to directly have `n = (n * q + r)%Z`, which could improve performance.

**Review by fajb — COMMENTED** [review]
Is there a reason for not including `ZifyN.v` and `ZifyNat.v` in `ZifyInst.v`?
(without setting `zify_convert_to_euclidean_division_equations_flag`)
Is it for backward compatibility?

**Comment by fajb** [comment]
> @fajb As #14086 is now merged, `zify` can be simplified.
> For example, consider `Goal forall (n:Z), (n mod n = 0)%Z.`. `zify` creates the assumption `n <> 0%Z -> n = (n * q + r)%Z`, where `n <> 0` is because of the old `mod` definition. Now, one can use the new `Z_div_mod_eq_full` to directly have `n = (n * q + r)%Z`, which could improve performance.

Sure. If @JasonGross agrees, go ahead! The culprit is `div_mod_to_equations_generalize` from `PreOmega.v`.

**Comment by mrhaandi** [comment]
> Is there a reason for not including `ZifyN.v` and `ZifyNat.v` in `ZifyInst.v`?
> (without setting `zify_convert_to_euclidean_division_equations_flag`)
> Is it for backward compatibility?

As you can imagine, `lia` becomes very powerful. I did not notice major performance regressions, but `ci` breaks too heavily. There is a lot of code that uses `lia` in a non-monotone way (i.e. if `lia` gets stronger, it breaks). So yes, backwards compatibility with anti-patterns.

**Comment by mrhaandi** [comment]
Since the new code is opt-in, for the current ci only minimal adjustments are required.
@silene could you take care of the required `flocq` adjustments (3 lines) for this (see https://github.com/mrhaandi/flocq/tree/N-Z-div-mod)?
@spitters could you look at https://github.com/coq-community/math-classes/pull/98 ?

**Comment by mrhaandi** [comment]
@fajb would you care to review the proposed additions? It would be great to have this until the deadline to be part of 8.14.
I also updated `zify` documentation to mention the new modules.

**Review by fajb — APPROVED** [review]
For me, this is good to go.

**Comment by jfehrle** [comment]
> As you can imagine, `lia` becomes very powerful. I did not notice major performance regressions, but `ci` breaks too heavily. There is a lot of code that uses `lia` in a non-monotone way (i.e. if `lia` gets stronger, it breaks). So yes, backwards compatibility with anti-patterns.

How hard would it be to fix the CI jobs that break (difficult or just tedious)?  The idea of keeping `lia` weaker than it could be seems unfortunate.

**Comment by mrhaandi** [comment]
> > As you can imagine, `lia` becomes very powerful. I did not notice major performance regressions, but `ci` breaks too heavily. There is a lot of code that uses `lia` in a non-monotone way (i.e. if `lia` gets stronger, it breaks). So yes, backwards compatibility with anti-patterns.
> 
> How hard would it be to fix the CI jobs that break (difficult or just tedious)? The idea of keeping `lia` weaker than it could be seems unfortunate.

From what I have looked at so far, quite tedious. However, I'm not sure whether including `div`, `mod`, `pow` translations for `nat` as always on is a good idea. It definitely deviates from integer linear arithmetic and may cause serious performance issues with larger linear goals, that by chance contain some of those operations.
In my opinion, it is best to have a dumb, fast default `lia` and a slow intelligent `lia` upon extra module inclusion (as of now).

**Comment by jfehrle** [comment]
Makes sense.  Perhaps you would tweak the documentation for `lia` to make the `zify` improvements easier to discover, perhaps something like
```Please see the description of :tacn:`zify` for information on extending `lia` to handle certain more general goals.```

**Comment by RalfJung** [comment]
It might also be a good idea to have another tactic that has the full preprocessing always enabled, so one doesn't have to wrangle with setting global hooks to get the more powerful behavior.

**Comment by mrhaandi** [comment]
> It might also be a good idea to have another tactic that has the full preprocessing always enabled, so one doesn't have to wrangle with setting global hooks to get the more powerful behavior.

@fajb wasn't `nia` supposed to be exactly this, i.e. goal massage + `lia`?

**Comment by fajb** [comment]
> @fajb wasn't `nia` supposed to be exactly this, i.e. goal massage + `lia`?

Nope. `nia` is more powerful for non-linear multiplications. It is using the same `zify`.

I support @RalfJung proposal of making easily available the best tactic.
Maybe, a `LiaExt.v` (or whatever name) doing all the required import?
PS: I do not think this is a blocking issue for the PR.

**Comment by RalfJung** [comment]
> Maybe, a LiaExt.v (or whatever name) doing all the required import?

I was imagining literally a separate tactic so that I could use `better_lia` and `lia` in the same proof. Or is there no reason to do that?

**Comment by fajb** [comment]
> > Maybe, a LiaExt.v (or whatever name) doing all the required import?
> 
> I was imagining literally a separate tactic so that I could use `better_lia` and `lia` in the same proof. Or is there no reason to do that?

Currently, I would not know how to do it. `zify` is using all the instances in the context. There is no notion of scope.

**Comment by mrhaandi** [comment]
> Currently, I would not know how to do it. `zify` is using all the instances in the context. There is no notion of scope.

Maybe it is unwise to put all instances into `typeclass_instances`. What about instead having individual hint databases for optional instances? This would allow for a natural notation `lia with hint_db` where `hint_db` is a (collection of) corresponding instance databases. Such an approach, as @RalfJung suggests, would allow for multiple `lia` variations to be uses side by side. Currently, once the optional instances are registered they are difficult (impossible?) to get rid of, leading to possible regressions.

**Comment by mrhaandi** [comment]
For your convenience (and for the deadline) I squashed the commits and rebased the PR on the current `master`.
Unfortunately the CI is stalled (and I have not enough rights to restart).
Also, there is still one minor overlay for `flocq`, could anyone besides @silene review and potentially merge it (3 line diff in the PR description) into `flocq`?
Otherwise, I'd consider this as good to go (further considerations should be considered in future PRs).

**Comment by silene** [comment]
> there is still one minor overlay for flocq

It has been merged.

**Comment by mrhaandi** [comment]
> > there is still one minor overlay for flocq
> 
> It has been merged.

Very nice. Now there are no more overlays. If ci agrees, this can go upstream cleanly.

**Comment by mrhaandi** [comment]
@fajb as everything seems in order, would you suggest a merge date?

**Comment by fajb** [comment]
Well done @mrhaandi !

**Pull request merged**
